[FSDP][optim_state_dict][1/N] Restructure _optim_state_dict to prepare the support of use_orig_param

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89981
* #89980
* #89900
* #89899
* __->__ #89898

**Motivation:**
Restructure some APIs in _optim_state_dict.py to allow better future extension, mostly for supporting use_orig_params. NO logic change in this PR.

fegin committed:
[FSDP][optim_state_dict][1/N] Restructure _optim_state_dict to prepare the support of use_orig_param

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict][1/N] Restructure _optim_state_dict to prepare the support of use_orig_param"




[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict][1/N] Restructure _optim_state_dict to prepare the support of use_orig_param"


**Motivation:**
Restructure some APIs in _optim_state_dict.py to allow better future extension, mostly for supporting use_orig_params. NO logic change in this PR.


[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -1071,51 +1070,104 @@ def _get_param_to_param_id_from_optim_input(
-        :class:`list`.
+def _map_param_id_to_optim_keys(
+    optim_state_dict: Dict[str, Any],
+    group: Optional[dist.ProcessGroup],
```
General comment: While refactoring optimizer state dict, it could be worthwhile to keep an eye out for process group usage (mainly looking for any accidental default process group usage). This is in preparation for HSDP, where every process group usage now should map to the intra-node process group from HSDP.

cc: @rohan-varma

awgu commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -1071,51 +1070,104 @@ def _get_param_to_param_id_from_optim_input(
+    return r0_param_id_to_optim_state_key, optim_state_key_to_param_id
+
+
+def _process_param_groups(
```
nit: I am wondering, should we call this `_unflatten_param_groups` to be more specific?

awgu reviewed (approved):
Thanks for this reorganization!

fegin committed:
Update on "[FSDP][optim_state_dict][1/N] Restructure _optim_state_dict to prepare the support of use_orig_param"


**Motivation:**
Restructure some APIs in _optim_state_dict.py to allow better future extension, mostly for supporting use_orig_params. NO logic change in this PR.


[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict][1/N] Restructure _optim_state_dict to prepare the support of use_orig_param"


**Motivation:**
Restructure some APIs in _optim_state_dict.py to allow better future extension, mostly for supporting use_orig_params. NO logic change in this PR.


[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict][1/N] Restructure _optim_state_dict to prepare the support of use_orig_param"


**Motivation:**
Restructure some APIs in _optim_state_dict.py to allow better future extension, mostly for supporting use_orig_params. NO logic change in this PR.


[ghstack-poisoned]
